File: Revario-Go/Day23/day23.go
package main

import (
	"fmt"
	"github.com/revario/advent_of_code_2016/Revario-Go/common"
	"strconv"
	"strings"
)

func main() {
	input := common.ReadFileByRow(`input.txt`)
	fmt.Printf("Part 1: %v, Part 2: %v", parseInstructions(input, loadInst(input), false), parseInstructions(input, loadInst(input), true))

}

func loadInst(input []string) map[int]string {
	inst := make(map[int]string)
	for i, v := range input {
		a := strings.Split(v, ` `)
		inst[i] = a[0]
	}
	return inst
}

func parseInstructions(input []string, instMap map[int]string, part2 bool) int {
	var reg = make(map[string]int)
	reg[`a`] = 7
	if part2 {
		reg[`a`] = 12
	}
	for n := 0; n < len(input); n++ {
		inst := strings.Split(input[n], ` `)
		switch instMap[n] {
		case `tgl`:
			var num int
			if i, err := strconv.Atoi(inst[1]); err == nil {
				num = n + i
			} else if i, ok := reg[inst[1]]; ok {
				num = n + i
			}
			switch instMap[num] {
			case `inc`:
				instMap[num] = `dec`
			case `dec`, `tgl`:
				instMap[num] = `inc`
			case `jnz`:
				instMap[num] = `cpy`
			case `cpy`:
				instMap[num] = `jnz`
			}

		case `inc`:
			if len(inst) != 2 {
				continue
			} else {
				reg[inst[1]]++
			}
		case `cpy`:

			if num, err := strconv.Atoi(inst[1]); err == nil {
				if _, err := strconv.Atoi(inst[2]); err != nil {
					reg[inst[2]] = num
				}
			} else {
				reg[inst[2]] = reg[inst[1]]
			}

		case `jnz`:
			if d, err := strconv.Atoi(inst[1]); err == nil {
				if d != 0 {
					if num, err := strconv.Atoi(inst[2]); err == nil {
						n += num - 1
					} else {
						n += reg[inst[2]] - 1
					}

				}

			} else {
				if reg[inst[1]] != 0 {
					if inst[2] == `-2` && ((instMap[n - 1] == `inc` && instMap[n - 2] == `dec`) || (instMap[n - 1] == `dec` && instMap[n - 2] == `inc`)) {
						if instMap[n - 1] == `inc` {

							ss := strings.Split(input[n - 1], ` `)

							reg[ss[1]] += reg[inst[1]]
						} else {
							ss := strings.Split(input[n - 2], ` `)

							reg[ss[1]] += reg[inst[1]]
						}
						reg[inst[1]] = 0

					} else if inst[2] == `-5` && inst[1] == `d` {

						reg[`a`] += reg[`b`] * reg[`d`]
						reg[`d`] = 0
					} else {

						if num, err := strconv.Atoi(inst[2]); err == nil {
							n += num - 1
						} else {
							n += reg[inst[2]] - 1
						}
					}

				}
			}


		case `dec`:
			reg[inst[1]]--
		default:
			panic(inst[0])
		}

	}
	return reg[`a`]
}

File: dlofstrom-python/Day23/day23.py

input = open('day23-input.txt','r')
input = [l.strip('\r\n').split(' ',1) for l in input]

registers = {'a':7, 'b':0, 'c':0, 'd':0}

def cpy(xy,i):
    x, y = xy.split(' ')

    #Detect multiply case
    if i+5 < len(input):
        if input[i+1][0] == 'inc' and input[i+2][0] == 'dec' and input[i+3][0] == 'jnz' and input[i+4][0] == 'dec' and input[i+5][0] == 'jnz':
            if input[i+2][1] == input[i+3][1][0] and input[i+4][1] == input[i+5][1][0]:
                if x.isdigit():
                    registers[input[i+1][1]] += int(x)*registers[input[i+4][1]]
                else:
                    registers[input[i+1][1]] += registers[x]*registers[input[i+4][1]]
                registers[input[i+2][1]] = 0
                registers[input[i+4][1]] = 0
                return 6

    #print 'cpy', xy    
    if y not in registers:
        return 1
    if x in registers:
        registers[y] = registers[x]
    else:
        registers[y] = int(x)
    return 1

def inc(x,i):
    #print 'inc', x
    registers[x] += 1
    return 1

def dec(x,i):
    #print 'dec', x
    registers[x] -= 1
    return 1

def jnz(xy,i):
    #print 'jnz', xy
    x, y = xy.split(' ')
    if x in registers:
        c = registers[x]
    else:
        c = int(x)
    if not c==0:
        if y in registers:
            return registers[y]
        else:
            return int(y)
    return 1

toggle_table = {'inc':'dec','dec':'inc','tgl':'inc','jnz':'cpy','cpy':'jnz'}

def tgl(x,i):
    #print 'tgl', x
    toggle = i + registers[x]
    if toggle >= len(input):
        return 1
    input[toggle][0] = toggle_table[input[toggle][0]]
    return 1

def assembunny_eval(reg):
    i = 0
    while True:
        #print i, input[i],registers
        i += eval(input[i][0]+'(\''+input[i][1]+'\',i)')
        if not i in range(len(input)):
            break
    return registers[reg]

print 'Part 1:', assembunny_eval('a')

input = open('day23-input.txt','r')
input = [l.strip('\r\n').split(' ',1) for l in input]

registers = {'a':12, 'b':0, 'c':0, 'd':0}

print 'Part 2:', assembunny_eval('a')
